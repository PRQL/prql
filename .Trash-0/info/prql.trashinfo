[Trash Info]
Path=/app/book/tests/prql
DeletionDate=2022-12-03T23:05:19
