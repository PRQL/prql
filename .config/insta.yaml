behavior:
  # force_update: true
review:
  # The default (true) has a small performance hit and never needed for us
  warn_undiscovered: false
  include_ignored: false
  include_hidden: false
