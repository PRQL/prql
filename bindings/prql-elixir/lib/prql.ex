defmodule PRQL do
  @moduledoc """
  Documentation for `PRQL`.

  This module provide Elixir bindings for [PRQL](https://prql-lang.org/).

  PRQL is a modern language for transforming data.
  """
  alias PRQL.Native.CompileOptions

  @type target() ::
          :generic
          | :mssql
          | :mysql
          | :postgres
          | :ansi
          | :bigquery
          | :clickhouse
          | :sqlite
          | :snowflake
  @type format_opt :: {:format, boolean()}
  @type signature_comment_opt :: {:signature_comment, boolean()}
  @type target_opt :: {:target, target()}
  @type compile_opts :: format_opt() | signature_comment_opt() | target_opt()

  @doc ~S"""
  Compile a `PRQL` query to `SQL` query.

  Returns generated SQL query on success in shape of `{:ok, sql}`. This SQL query can
  be safely fed into a SQL driver to get the output.

  On error, `{:error, reason}` is returned where `reason` is a `JSON` string.
  Use any `JSON` encoder library to encode it.

  ## Options

    * `:target` - Dialect used for generate SQL. Accepted values are
    `:generic`, `:mssql`, `:mysql`, `:postgres`, `:ansi`, `:bigquery`,
    `:clickhouse`, `:sqlite`, `:snowflake`

    * `:format` - Formats the output, defaults to `true`

    * `signature_comment` - Set the signature comment generated by PRQL, defaults to `true`


  ## Examples

  Using default `Generic` target:
      iex> PRQL.compile("from customers", signature_comment: false)
      {:ok, "SELECT\n  *\nFROM\n  customers\n"}


  Using `MSSQL` target:
      iex> PRQL.compile("from customers\ntake 10", target: :mssql, signature_comment: false)
      {:ok, "SELECT\n  TOP (10) *\nFROM\n  customers\n"}
  """
  @spec compile(binary(), [compile_opts()]) :: {:ok, binary()} | {:error, binary()}
  def compile(prql_query, opts \\ []) when is_binary(prql_query) and is_list(opts) do
    PRQL.Native.compile(prql_query, struct(CompileOptions, opts))
  end

  @doc """
  The same as `compile/2` but raises `PRQL.PRQLError` exception in case of error
  """
  @spec compile!(binary(), [compile_opts()]) :: binary()
  def compile!(prql_query, opts \\ []) do
    case compile(prql_query, opts) do
      {:ok, result} -> result
      {:error, reason} -> raise PRQL.PRQLError, reason
    end
  end

  @doc """
  PRQL to PL AST
  """
  @spec prql_to_pl(binary()) :: {:ok, binary()} | {:error, binary()}
  def prql_to_pl(prql_query) when is_binary(prql_query) do
    PRQL.Native.prql_to_pl(prql_query)
  end

  @doc """
  The same as `prql_to_pl/1` but raises `PRQL.PRQLError` exception in case of error
  """
  @spec prql_to_pl!(binary()) :: binary()
  def prql_to_pl!(prql_query) when is_binary(prql_query) do
    case prql_to_pl(prql_query) do
      {:ok, result} -> result
      {:error, reason} -> raise PRQL.PRQLError, reason
    end
  end

  @doc """
  PL AST to RQ
  """
  @spec pl_to_rq(binary()) :: {:ok, binary()} | {:error, binary()}
  def pl_to_rq(pl_json) do
    PRQL.Native.pl_to_rq(pl_json)
  end

  @doc """
  The same as `pl_to_rq/1` but raises `PRQL.PRQLError` exception in case of error
  """
  @spec pl_to_rq!(binary()) :: binary()
  def pl_to_rq!(pl_json) do
    case pl_to_rq(pl_json) do
      {:ok, result} -> result
      {:error, reason} -> raise PRQL.PRQLError, reason
    end
  end

  @doc """
  RQ to SQL
  """
  @spec rq_to_sql(binary()) :: {:ok, binary()} | {:error, binary()}
  def rq_to_sql(rq_json) do
    PRQL.Native.rq_to_sql(rq_json)
  end

  @doc """
  The same as `rq_to_sql/1` but raises `PRQL.PRQLError` exception in case of error
  """
  @spec rq_to_sql!(binary()) :: binary()
  def rq_to_sql!(rq_json) do
    case rq_to_sql(rq_json) do
      {:ok, result} -> result
      {:error, reason} -> raise PRQL.PRQLError, reason
    end
  end
end
