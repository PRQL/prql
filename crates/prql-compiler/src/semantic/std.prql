# The PRQL standard library defines the following functions and transforms.
# The definitions are whitespace insensitive, and have this form:
#
# ```
# let my_func = param1 param2 ...  -> <return_type> body_expr
# ```
#
# Where:
#   * `my_func` is the name of the function
#   * `param1` is the first parameter optionally followed by a type in "< ... >"
#   * `param2` etc. follow the same pattern as param1
#   * `<return_type>` is the type of result wrapped in "< ... >"
#   * `body_expr` defines the function body that creates the result.
#     It can be PRQL code or `internal ...` to indicate internal compiler code.

# Operators

let mul = left right -> <int || float> internal std.mul
let div_i = left right -> <int || float> internal std.div_i
let div_f = left right -> <int || float> internal std.div_f
let mod = left right -> <int || float> internal std.mod
let add = left<int || float || timestamp || date> right<int || float || timestamp || date> -> <int || float || timestamp || date> internal std.add
let sub = left<int || float || timestamp || date> right<int || float || timestamp || date> -> <int || float || timestamp || date> internal std.sub
let eq = left right -> <bool> internal std.eq
let ne = left right -> <bool> internal std.ne
let gt = left right -> <bool> internal std.gt
let lt = left right -> <bool> internal std.lt
let gte = left right -> <bool> internal std.gte
let lte = left right -> <bool> internal std.lte
let and = left<bool> right<bool> -> <bool> internal std.and
let or = left<bool> right<bool> -> <bool> internal std.or
let coalesce = left right -> internal std.coalesce
let regex_search = text pattern -> <bool> internal std.regex_search

let neg = expr<int || float> -> <int || float> internal std.neg
let not = expr<bool> -> <bool> internal std.not

# Types

## Type primitives
type int
type float
type bool
type text
type date
type time
type timestamp
type `func`

## Generic array
# TODO: an array of anything, not just nulls
type array = [null]

## Scalar
type scalar = int || float || bool || text || date || time || timestamp || null
type tuple_of_scalars = {scalar..}

## Relation (an array of tuples)
type relation = [tuple_of_scalars]

## Transform
type transform = (func relation -> relation)

# Functions

## Relational transforms
let from = func
  `default_db.source` <relation>
  -> <relation> internal from

let select = func
  columns <scalar || tuple_of_scalars>
  tbl <relation>
  -> <relation> internal select

let filter = func
  condition <bool>
  tbl <relation>
  -> <relation> internal filter

let derive = func
  columns <scalar || tuple_of_scalars>
  tbl <relation>
  -> <relation> internal derive

let aggregate = func
  columns <scalar || tuple_of_scalars>
  tbl <relation>
  -> <relation> internal aggregate

let sort = func
  by <scalar || tuple_of_scalars>
  tbl <relation>
  -> <relation> internal sort

let take = func
  expr <scalar>
  tbl <relation>
  -> <relation> internal take

let join = func
  `default_db.with` <relation>
  condition <bool>
  `noresolve.side`:inner
  tbl <relation>
  -> <relation> internal join

let group = func
  by<scalar || tuple_of_scalars>
  pipeline <transform>
  tbl <relation>
  -> <relation> internal group

let window = func
  rows:0..0
  range:0..0
  expanding <bool>:false
  rolling <int>:0
  pipeline <transform>
  tbl <relation>
  -> <relation> internal window

let noop = x -> x

let append = `default_db.bottom`<relation> top<relation> -> <relation> internal append
let intersect = `default_db.bottom`<relation> top<relation> -> <relation> (
  noop t = top
  join (noop b = bottom) (tuple_every (tuple_map _eq (tuple_zip t.* b.*)))
  select t.*
)
let remove = `default_db.bottom`<relation> top<relation> -> <relation> (
  noop t = top
  join side:left (noop b = bottom) (tuple_every (tuple_map _eq (tuple_zip t.* b.*)))
  filter (tuple_every (tuple_map _is_null b.*))
  select t.*
)
let loop = func
  pipeline <transform>
  top <relation>
  -> <relation> internal loop

## Aggregate functions
# These return either a scalar when used within `aggregate`, or a column when used anywhere else.

let min = column <array> -> <int || float || null> internal std.min

let max = column <array> -> <int || float || null> internal std.max

let sum = column <array> -> <int || float> internal std.sum

let average = column <array> -> <float || null> internal std.average

let stddev = column <array> -> <float || null> internal std.stddev

let every = column <array> -> <bool> internal std.every

let any = column <array> -> <bool> internal std.any

let concat_array = column <array> -> <text> internal std.concat_array

# Counts number of items in the column.
# Note that the count will include null values.
let count = column<array> -> <int> internal std.count

# Deprecated in favour of filterning input to the [std.count] function (not yet implemented).
@{deprecated}
let count_distinct = column <array> -> internal std.count_distinct

## Window functions
let lag =   offset <int>    column <array> -> internal std.lag
let lead =  offset <int>    column <array> -> internal std.lead
let first      = column <array> -> internal std.first
let last       = column <array> -> internal std.last
let rank       = column <array> -> internal std.rank
let rank_dense = column <array> -> internal std.rank_dense
let row_number = column <array> -> internal std.row_number

## Misc functions
let round = n_digits column -> <scalar> internal std.round
let as = `noresolve.type` column -> <scalar> internal std.as
let in = pattern value -> <bool> internal in

## Tuple functions
let tuple_every = func list -> <bool> internal tuple_every
let tuple_map = func fn <func> list -> internal tuple_map
let tuple_zip = func a b -> internal tuple_zip
let _eq = func a -> internal _eq
let _is_null = func a -> _param.a == null

## Misc
let from_text = input<text> `noresolve.format`:csv -> <relation> internal from_text

## String functions
let lower = column -> <text> internal std.lower
let upper = column -> <text> internal std.upper

## File-reading functions, primarily for DuckDB
let read_parquet = source<text> -> <relation> internal std.read_parquet
let read_csv = source<text> -> <relation> internal std.read_csv
