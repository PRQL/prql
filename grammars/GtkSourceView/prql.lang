<?xml version="1.0" encoding="UTF-8"?>
<!--

 Copyright (C) 2024 The PRQL Project

 https://prql-lang.org/
 https://github.com/PRQL/prql

-->
<language id="prql" name="PRQL" version="2.0" _section="Source">
  <metadata>
    <property name="mimetypes">application/x.prql;application/prs.prql;application/vnd.prql</property>
    <property name="globs">*.prql</property>
    <property name="line-comment-start">#</property>
    <property name="suggested-suffix">.prql</property>
  </metadata>

  <styles>
    <style id="base-n-integer"  name="Base-N number"     map-to="def:base-n-integer"/>
    <style id="boolean"         name="Boolean value"     map-to="def:boolean"/>
    <style id="built-in-object" name="Built-in object"   map-to="def:builtin"/>
    <style id="comment"         name="Comment"           map-to="def:comment"/>
    <style id="declaration"     name="Declarations"      map-to="def:type"/>
    <style id="escaped-char"    name="Escaped Character" map-to="def:special-char"/>
    <style id="floating-point"  name="Floating Point"    map-to="def:floating-point"/>
    <style id="string"          name="String"            map-to="def:string"/>
    <style id="keyword"         name="Keyword"           map-to="def:keyword"/>
    <style id="function"        name="Function"          map-to="def:function"/>
    <style id="decimal"         name="Decimal"           map-to="def:decimal"/>
    <style id="type"            name="Data Type"         map-to="def:type"/>
    <style id="f-string-curly-braces" name="f-string curly braces" map-to="def:special-char"/>
    <style id="unicode-bdi"    name="Unicode BDI"       map-to="def:error"/>
  </styles>

  <definitions>

    <context id="boolean" style-ref="boolean">
      <keyword>false</keyword>
      <keyword>true</keyword>
    </context>

    <context id="declarations" style-ref="declaration">
      <keyword>let</keyword>
    </context>

    <context id="unicode-bdi" style-ref="unicode-bdi">
      <match>(\u202A|\u202B|\u202D|\u202E|\u2066|\u2067|\u2068|\u202C|\u2069)</match>
    </context>

    <context id="illegal-string" style-ref="unicode-bdi">
      <match>\b[^frs:][\"\']</match>
    </context>

    <context id="psql-datatypes" style-ref="type">
      <keyword>bool</keyword>
      <keyword>float</keyword>
      <keyword>int</keyword>
      <keyword>int8</keyword>
      <keyword>int16</keyword>
      <keyword>int32</keyword>
      <keyword>int64</keyword>
      <keyword>int128</keyword>
      <keyword>text</keyword>
      <keyword>date</keyword>
      <keyword>time</keyword>
      <keyword>timestamp</keyword>
    </context>

    <context id="integer-literals" style-ref="decimal">
      <match>\b[0-9_]+(?![Ee][\+\-]?[0-9]*)\b</match>
    </context>

    <context id="number-literals" style-ref="floating-point">
      <match>(?&lt;![\w\.])(([0-9]+\.[0-9]*|\.[0-9]+)([Ee][\+\-]?[0-9]*)?|([0-9]+[Ee][\+\-]?[0-9]*))(?![\w\.])</match>
    </context>

    <context id="null" style-ref="decimal">
      <keyword>null</keyword>
    </context>

    <context id="block-comment" style-ref="comment" class="comment" end-at-line-end="true" class-disabled="no-spell-check">
      <start>#!</start>
      <include>
        <context ref="def:in-line-comment"/>
      </include>
    </context>

    <context id="line-comment" style-ref="comment" end-at-line-end="true" class="comment" class-disabled="no-spell-check">
      <start>#</start>
      <include>
        <context ref="def:in-line-comment"/>
      </include>
    </context>

    <context id="built-in-object" style-ref="built-in-object">
      <keyword>date</keyword>
      <keyword>math</keyword>
      <keyword>prql</keyword>
    </context>

    <!-- https://prql-lang.org/book/reference/stdlib/ -->
    <context id="aggregate-functions" style-ref="function">
      <keyword>any</keyword>
      <keyword>average</keyword>
      <keyword>concat_array</keyword>
      <keyword>count</keyword>
      <keyword>every</keyword>
      <keyword>max|min</keyword>
      <keyword>stddev</keyword>
      <keyword>sum</keyword>
    </context>

    <context id="file-reading-functions" style-ref="function">
      <keyword>read_csv</keyword>
      <keyword>read_parquet</keyword>
    </context>

    <context id="list-functions" style-ref="function">
      <keyword>all</keyword>
      <keyword>map</keyword>
      <keyword>zip</keyword>
      <keyword>_eq</keyword>
      <keyword>_is_null</keyword>
    </context>

    <context id="misc-functions" style-ref="function">
      <keyword>from_text</keyword>
    </context>

    <context id="text-functions" style-ref="function">
      <keyword>contains</keyword>
      <keyword>ends_with</keyword>
      <keyword>extract</keyword>
      <keyword>length</keyword>
      <keyword>lower</keyword>
      <keyword>ltrim</keyword>
      <keyword>replace</keyword>
      <keyword>rtrim</keyword>
      <keyword>starts_with</keyword>
      <keyword>trim</keyword>
      <keyword>upper</keyword>
    </context>

    <context id="window-functions" style-ref="function">
      <keyword>lag|lead</keyword>
      <keyword>first|last</keyword>
      <keyword>rank</keyword>
      <keyword>rank_dense</keyword>
      <keyword>row_number</keyword>
    </context>

    <context id="transform-type-definitions" style-ref="function">
      <keyword>aggregate</keyword>
      <keyword>derive</keyword>
      <keyword>filter</keyword>
      <keyword>from</keyword>
      <keyword>group</keyword>
      <keyword>join</keyword>
      <keyword>select</keyword>
      <keyword>sort</keyword>
      <keyword>take</keyword>
      <keyword>window</keyword>
    </context>

    <context id="date-functions" style-ref="function">
      <keyword>to_text</keyword>
    </context>

    <context id="math-functions" style-ref="function">
      <keyword>abs</keyword>
      <keyword>acos</keyword>
      <keyword>asin</keyword>
      <keyword>atan</keyword>
      <keyword>ceil</keyword>
      <keyword>cos</keyword>
      <keyword>degrees</keyword>
      <keyword>exp</keyword>
      <keyword>floor</keyword>
      <keyword>ln</keyword>
      <keyword>log</keyword>
      <keyword>log10</keyword>
      <keyword>pi</keyword>
      <keyword>pow</keyword>
      <keyword>radians</keyword>
      <keyword>round</keyword>
      <keyword>sin</keyword>
      <keyword>sqrt</keyword>
      <keyword>tan</keyword>
    </context>

    <define-regex id="identifier" extended="true">
      (?&gt; (?: _ | \%{def:unicode-xid-start} ) \%{def:unicode-xid-continue}* )
    </define-regex>
    <define-regex id="number">[1-9][0-9]*</define-regex>

    <!-- https://prql-lang.org/book/reference/syntax/strings.html -->
    <context id="escaped-char" style-ref="escaped-char" extend-parent="true">
      <match extended="true">
        \\(                   # leading backslash
        [\\'"bfnrt]         | # single escaped char
        u{[0-9A-Fa-f]{1,6}} | # \u{hhhhhh} - unicode character
        x[0-9A-Fa-f]{1,2}   | # \xhh - character with hex value hh
        )
      </match>
    </context>

    <context id="curly-braces" extend-parent="true">
      <start>\{</start>
      <end>\}</end>
      <include>
        <context ref="prql"/>
        <context ref="curly-braces"/>
      </include>
    </context>

    <context id="f-string-curly-braces" extend-parent="false" class-disabled="string">
      <start>(\{)</start>
      <end>(\})</end>
      <include>
        <context ref="prql"/>
        <context ref="curly-braces"/>
        <context sub-pattern="1" where="start" style-ref="f-string-curly-braces"/>
        <context sub-pattern="1" where="end" style-ref="f-string-curly-braces"/>
      </include>
    </context>

    <context id="escaped-curly-brace" style-ref="escaped-char" extend-parent="true">
      <match>\{\{</match>
    </context>

    <context id="double-quoted-string" style-ref="string" end-at-line-end="true" class="string" class-disabled="no-spell-check">
      <start>"</start>
      <end>"</end>
      <include>
        <context ref="escaped-char"/>
        <context ref="def:line-continue"/>
      </include>
    </context>

    <context id="single-quoted-string" style-ref="string" end-at-line-end="true" class="string" class-disabled="no-spell-check">
      <start>'</start>
      <end>'</end>
      <include>
        <context ref="escaped-char"/>
        <context ref="def:line-continue"/>
      </include>
    </context>

    <context id="double-quoted-triple-string" style-ref="string" class="string" class-disabled="no-spell-check">
      <start>"""</start>
      <end>"""</end>
      <include>
        <context ref="escaped-char"/>
        <context ref="def:line-continue"/>
      </include>
    </context>

    <context id="single-quoted-triple-string" style-ref="string" class="string" class-disabled="no-spell-check">
      <start>'''</start>
      <end>'''</end>
      <include>
        <context ref="escaped-char"/>
        <context ref="def:line-continue"/>
      </include>
    </context>

    <context id="double-quoted-f-string" end-at-line-end="true" class="string" class-disabled="no-spell-check">
      <start>(f")</start>
      <end>(")</end>
      <include>
        <context ref="escaped-curly-brace"/>
        <context ref="f-string-curly-braces"/>
        <context ref="escaped-char"/>
        <context ref="def:line-continue"/>
        <context style-ref="string" extend-parent="false" class="string">
          <match>.</match>
        </context>
        <context sub-pattern="1" where="start" style-ref="string"/>
        <context sub-pattern="1" where="end" style-ref="string"/>
      </include>
    </context>

    <context id="single-quoted-f-string" end-at-line-end="true" class="string" class-disabled="no-spell-check">
      <start>(f')</start>
      <end>(')</end>
      <include>
        <context ref="escaped-curly-brace"/>
        <context ref="f-string-curly-braces"/>
        <context ref="escaped-char"/>
        <context ref="def:line-continue"/>
        <context style-ref="string" extend-parent="false" class="string">
          <match>.</match>
        </context>
        <context sub-pattern="1" where="start" style-ref="string"/>
        <context sub-pattern="1" where="end" style-ref="string"/>
      </include>
    </context>

    <context id="double-quoted-r-string" style-ref="string" end-at-line-end="true" class="string" class-disabled="no-spell-check">
      <start>r"</start>
      <end>"</end>
      <include>
        <context ref="def:line-continue"/>
      </include>
    </context>

    <context id="single-quoted-r-string" style-ref="string" end-at-line-end="true" class="string" class-disabled="no-spell-check">
      <start>r'</start>
      <end>r'</end>
      <include>
        <context ref="def:line-continue"/>
      </include>
    </context>

    <context id="double-quoted-s-string" style-ref="string" end-at-line-end="true" class="string" class-disabled="no-spell-check">
      <start>s"</start>
      <end>"</end>
      <include>
        <context ref="def:line-continue"/>
      </include>
    </context>

    <context id="single-quoted-s-string" style-ref="string" end-at-line-end="true" class="string" class-disabled="no-spell-check">
      <start>s'</start>
      <end>'</end>
      <include>
        <context ref="def:line-continue"/>
      </include>
    </context>

    <context id="dimension" style-ref="floating-point">
      <prefix>\%{number}</prefix>
      <keyword>microseconds</keyword>
      <keyword>milliseconds</keyword>
      <keyword>seconds</keyword>
      <keyword>minutes</keyword>
      <keyword>hours</keyword>
      <keyword>days</keyword>
      <keyword>weeks</keyword>
      <keyword>months</keyword>
      <keyword>years</keyword>
    </context>

    <context id="prql" class="no-spell-check">
      <include>
        <context ref="boolean"/>

        <context id="binary" style-ref="base-n-integer">
          <match>(?&lt;![\w\.])0[bB](_?[0-1])+(?![\w\.])</match>
        </context>

        <context id="octal" style-ref="base-n-integer">
          <match>(?&lt;![\w\.])0[oO](_?[0-7])+(?![\w\.])</match>
        </context>

        <context id="hex" style-ref="base-n-integer">
          <match>(?&lt;![\w\.])0[xX](_?[0-9A-Fa-f])+(?![\w\.])</match>
        </context>

        <context ref="declarations"/>
        <context ref="dimension"/>
        <context ref="double-quoted-triple-string"/>
        <context ref="single-quoted-triple-string"/>
        <context ref="double-quoted-f-string"/>
        <context ref="single-quoted-f-string"/>
        <context ref="double-quoted-r-string"/>
        <context ref="single-quoted-r-string"/>
        <context ref="double-quoted-s-string"/>
        <context ref="single-quoted-s-string"/>
        <context ref="double-quoted-string"/>
        <context ref="single-quoted-string"/>
        <context ref="psql-datatypes"/>
        <context ref="number-literals"/>
        <context ref="integer-literals"/>
        <context ref="null"/>
        <context ref="block-comment"/>
        <context ref="line-comment"/>
        <context ref="built-in-object"/>
        <context ref="aggregate-functions"/>
        <context ref="date-functions"/>
        <context ref="file-reading-functions"/>
        <context ref="list-functions"/>
        <context ref="math-functions"/>
        <context ref="misc-functions"/>
        <context ref="text-functions"/>
        <context ref="transform-type-definitions"/>
        <context ref="unicode-bdi"/>
        <context ref="illegal-string"/>
        <context ref="window-functions"/>
      </include>
    </context>

  </definitions>
</language>
