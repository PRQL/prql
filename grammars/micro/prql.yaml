# PRQL syntax highlighting file for micro - https://micro-editor.github.io/
# https://prql-lang.org/
# https://github.com/PRQL/prql
filetype: prql

detect:
  filename: "\\.prql$"

rules:
  - statement: "\\b(let|module)\\b"

    # Types
  - type: "\\b(u?int(8|16|32|64)?|float(32|64)|bool|text|date|time|timestamp)\\b"
  - type.keyword: "\\b(enum)\\b"
  - constant.bool: "\\b(true|false|null|this|that)\\b"

    # Built-in functions
  - identifier: "\\b(abs|floor|ceil|pi|exp|ln|log10|log|sqrt|degrees|radians|cos|acos|sin|asin|tan|atan|pow|round)\\b" # Math module
  - identifier: "\\b(min|max|sum|average|stddev|all|any|concat_array|count)\\b" # Aggregate functions
  - identifier: "\\b(lag|lead|first|last|rank|rank_dense|row_number)\\b" # Window functions
  - identifier: "\\b(tuple_every|tuple_map|tuple_zip|_eq|_is_null)\\b" # Tuple functions
  - identifier: "\\b(as|in|from_text)\\b" # Misc
  - identifier: "\\b(lower|upper|ltrim|rtrim|trim|length|extract|replace|starts_with|contains|ends_with)\\b" # Text module
  - identifier: "\\b(to_text)\\b" # Date module
  - identifier: "\\b(read_parquet|read_csv)\\b" # File-reading functions

    # Modules
  - identifier.class: "\\b(math|text|date|prql)\\b"

    # Transforms
  - statement: "\\b(aggregate|derive|filter|from|group|join|select|sort|take|window)\\b"

    # Operators
  - symbol.operator: "([~^.:;,+*|=!\\%@]|<|>|/|-|&)"

    # Brackets
  - symbol.brackets: "(\\{|\\})"
  - symbol.brackets: "(\\(|\\))"
  - symbol.brackets: "(\\[|\\])"

    # Numbers and strings
  - constant.number: "\\b[0-9](_?[0-9])*(\\.([0-9](_?[0-9])*)?)?(e[0-9](_?[0-9])*)?\\b" # decimal
  - constant.number: "\\b0b(_?[01])+\\b" # bin
  - constant.number: "\\b0o(_?[0-7])+\\b" # oct
  - constant.number: "\\b0x(_?[0-9a-fA-F])+\\b" # hex
  - constant: "\\b[0-9]+(years|months|weeks|days|hours|minutes|seconds|milliseconds|microseconds)\\b"

  - constant.string:
      start: '[frs]?"'
      end: '"'
      skip: "\\\\."
      rules:
        - constant.specialChar: "\\\\[bfnrt'\\\"\\\\]"
        - constant.specialChar: "\\\\([0-7]{3}|x[A-Fa-f0-9]{2}|u{[0-9A-Fa-f]{1,6}})"

  - constant.string:
      start: "[frs]?'"
      end: "'"
      skip: "\\\\."
      rules:
        - constant.specialChar: "\\\\."
        - constant.specialChar: "\\\\([0-7]{3}|x[A-Fa-f0-9]{2}|u{[0-9A-Fa-f]{1,6}})"

  - comment:
      start: "#"
      end: "$"
      rules:
        - todo: "(TODO|FIXME|NOTE):?"

    # Decorators
  - preproc: "@\\{([a-z]+(=[a-z0-9]+,?)?)*\\}"
