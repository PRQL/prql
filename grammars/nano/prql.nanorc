## Syntax highlighting for PRQL.

syntax python "\.prql$"
magic "PRQL script"
comment "#"

# Types.
color green "\<(int(8|16|32|64|128)?|float(32|64)|bool|text|date|time|timestamp)\>"

# Keywords.
color yellow "\<let|module|prql\>"

# Transforms.
color brightcyan "\<(aggregate|derive|filter|from|group|join|select|sort|take|window)\>"

# Special values.
color brightmagenta "\<(false|null|true|this|that)\>"

# Decorators.
color cyan start="@\{" end="\}"

# Mono-quoted strings.
color brightgreen "[frs]?'([^'\]|\\.)*'|[frs]?"([^"\]|\\.)*"|'''|""""
color normal "'''|""""
# Comments.
color gray "(^|[[:blank:]])#.*"
# Triple-quoted strings.
color brightgreen start="[frs]?'''([^'),]|$)" end="(^|[^(\])'''"
color brightgreen start="[frs]?"""([^"),]|$)" end="(^|[^(\])""""

# Backslash escapes.
color lime "\\($|[\'"bfnrt]|[0-3]?[0-7]?[0-7]|x[[:xdigit:]]{2})"
color lime "\\(N\{[[:alpha:]]+\}|u\{[[:xdigit:]]{1,6}\})"

# Reminders.
color brightwhite,yellow "\<(FIXME|TODO)\>"

# Trailing whitespace.
color ,green "[[:space:]]+$"
