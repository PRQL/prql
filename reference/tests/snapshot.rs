/// This test:
/// - Extracts PRQL code blocks into the `examples` path.
/// - Converts them to SQL using insta, raising an error if there's a diff.
/// - Replaces the PRQL code block with a comparison table.
//
// Overall, this is bad quality code that I shouldn't have written. It also
// took a long time to write. The intention was reasonable â€” have a version of
// the SQL that's committed into the repo, and join our tests with our docs.
//
// We don't use a book preprocessor, because we want to the results committed
// into the repository, so we can see if anything changes (I think this
// dimension is quite important.)
//
// Possibly we should be using something like pandoc /
// https://github.com/gpoore/codebraid / which would run the transformation for
// us. They introduce a bunch of non-rust dependencies, which is not ideal, but
// passable. They don't let us customize our formatting (e.g. in a table).
//
// Overall, this feels like overly custom code for quite a general problem, even
// if our preferences are slightly different.
use anyhow::{bail, Result};
use globset::Glob;
use insta::{assert_snapshot, glob};
use prql::*;
use pulldown_cmark::{CodeBlockKind, Event, Parser, Tag};
use std::fs;
use std::path::Path;
use walkdir::WalkDir;

#[test]
fn run_examples() -> Result<()> {
    write_reference_examples()?;
    run_reference_examples()?;

    Ok(())
}

/// Extract reference examples from the PRQL docs and write them to the
/// `tests/examples` path, one in each file.
// We could alternatively have used something like
// https://github.com/earldouglas/codedown, but it's not much code, and it
// requires no dependencies.
fn write_reference_examples() -> Result<()> {
    let glob = Glob::new("**/*.md")?.compile_matcher();

    WalkDir::new(Path::new("./src/"))
        .into_iter()
        .flatten()
        .filter(|x| glob.is_match(x.path()))
        .try_for_each(|dir_entry| {
            let text = fs::read_to_string(dir_entry.path())?;
            let mut parser = Parser::new(&text);
            let mut examples = vec![];
            while let Some(event) = parser.next() {
                match event.clone() {
                    // At the start of a PRQL code block, push the _next_ item.
                    Event::Start(Tag::CodeBlock(CodeBlockKind::Fenced(lang)))
                        if lang == "prql".into() =>
                    {
                        if let Some(Event::Text(text)) = parser.next() {
                            examples.push(text);
                        } else {
                            bail!("Expected text after PRQL code block");
                        }
                    }
                    _ => {}
                }
            }

            // Write each one to a new file.
            examples.iter().enumerate().try_for_each(|(i, example)| {
                let file_relative = &dir_entry
                    .path()
                    .strip_prefix("./src/")?
                    .to_str()
                    .unwrap()
                    .trim_end_matches(".md");
                let prql_path = format!("tests/examples/{file_relative}-{i}.prql");

                fs::create_dir_all(Path::new(&prql_path).parent().unwrap())?;
                fs::write(prql_path, example.to_string())?;

                Ok::<(), anyhow::Error>(())
            })?;
            Ok(())
        })?;

    Ok(())
}

/// Snapshot the output of each example.
fn run_reference_examples() -> Result<()> {
    glob!("examples/**/*.prql", |path| {
        let prql = fs::read_to_string(path).unwrap();

        if prql.contains("skip_test") {
            return;
        }

        let sql = compile(&prql).unwrap_or_else(|e| format!("Failed to compile `{prql}`; {e}"));
        assert_snapshot!(sql);
    });
    Ok(())
}
