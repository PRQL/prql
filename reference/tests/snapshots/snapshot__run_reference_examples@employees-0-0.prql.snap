---
source: reference/tests/snapshot.rs
expression: sql
input_file: reference/tests/examples/examples/employees-0-0.prql
---
WITH table_0 AS (
  SELECT
    emp_no,
    AVG(salary) AS emp_salary
  FROM
    salaries
  GROUP BY
    emp_no
),
table_1 AS (
  SELECT
    dept_emp.dept_no,
    t.title,
    AVG(emp_salary) AS avg_salary
  FROM
    table_0
    JOIN titles AS t USING(emp_no)
    LEFT JOIN dept_emp USING(emp_no)
  GROUP BY
    dept_emp.dept_no,
    t.title
)
SELECT
  dept_name,
  title,
  avg_salary
FROM
  table_1
  JOIN departments USING(dept_no)
